### DISCUSSION TITLE
Parallelizing github actions

### DISCUSSION THREAD

**Discussion by jordandsullivan** [unitaryfund/mitiq#2308] [Q&A] [answered]
Apologies if this is a duplicate, but I wanted to know if we had decided on a tool for parallelizing the docs build part of the github actions. My understanding is that it's actually running all the notebooks that takes a long time (as pointed out by @natestemen in [#2247](https://github.com/unitaryfund/mitiq/issues/2247#issuecomment-2064816465), these example notebooks are essentially our integration testing at present).

Locally I have started using this tool https://pypi.org/project/pytest-xdist/ when I run pytest, which has an automatic mode to parallelize across as many CPUs as are available.

**Comment by natestemen** [comment]
### Docs

We have https://github.com/unitaryfund/mitiq/pull/2235 opened from @cosenal which is currently blocked by the sphinx theme that we use. Alessandro has done some testing for the pydata-sphinx-theme and found our docs look okay even when built in parallel. See the comment [here](https://github.com/pydata/pydata-sphinx-theme/issues/1643#issuecomment-2023366648).

### Tests

When running `make test` the following command is run:
https://github.com/unitaryfund/mitiq/blob/e76ef0898b964c928b9a3bd8dddf698e51a7f7aa/Makefile#L63

In particular the `-n auto` option is provided by the `pytest-xdist` package and is described in `pytest --help`:
```
  -n numprocesses, --numprocesses=numprocesses
                        Shortcut for '--dist=load --tx=NUM*popen'. With 'auto',
                        attempt to detect physical CPU count. With 'logical',
                        detect logical CPU count. If physical CPU count cannot
                        be found, falls back to logical count. This will be 0
                        when used with --pdb.
```

You can see `pytest-xdist` is already in our [`dev_requirements.txt`](https://github.com/unitaryfund/mitiq/blob/e76ef0898b964c928b9a3bd8dddf698e51a7f7aa/dev_requirements.txt#L13) 😃. It helps the tests run so much faster locally! Usually when I'm just testing the changes I've made I run a single test file, or a series of files in a single module, but when running more than that I almost always use `-n auto`.

**Reply by jordandsullivan** [reply, marked as answer]
I see. So is `make test` already being run in parallel on the [Github runner](https://docs.github.com/en/actions/using-github-hosted-runners/about-github-hosted-runners/about-github-hosted-runners).

**Reply by jordandsullivan** [reply]
Reference to the place in the Makefile where this is written:
https://github.com/unitaryfund/mitiq/blob/6ff6c0841e2e24a32917b235b4ffe9d20256c678/Makefile#L61-L67
